addons/libs/images.lua

```diff
@@ -407,6 +426,9 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
             if meta[dragged.image].root_settings then
                 config.save(meta[dragged.image].root_settings)
             end
+            if x == clicked.x and y == clicked.y then
```

Comment:
I wonder if we need to be this exact with the click. Maybe a small tolerance is fine? Also maybe wrap it into a `settings.clickable` option?

Either way, this is currently under the `if dragged` condition, this should be moved outside in an analogous `if clicked` condition.

Reply:
Would an average delta of 2 be an ideal tolerance? _x delta / y delta = "average delta"_. Using my sample addon, I see that most of my gradual drags have an average delta of 0.5.

Reply:
Also, I placed it in the `if dragged` condition because, assuming A) all draggable objects would go into that if branc and B) undraggable objects don't need to know where they were clicked, it would be unecessary overhead to include it outside the branch. Undraggable objects, since they can't be moved, can just fire the left_click event the moment they are fired and never worry, or get/set, the clicked.x and clicked.y values, right?

Reply:
> Using my sample addon, I see that most of my gradual drags have an average delta of 0.5.

Huh? 0.5? How are you measuring delta, by pixel value? If so, an average may be the wrong tool. I'd use something like the 90th percentile. And even then I'd add something like a flat amount to that, to cover people who may be less accurate than you, in general.

There might be an overall better solution to this, but not sure. You seem to want to combine being draggable with being clickable, which makes this problem unnecessarily hard. Ideally I would solve those two problems separately and assume that being clickable and being draggable are mutually exclusive. The more I think about the more I think we should go that route. Would make this easy to solve - if `clicked` is not `nil` just check on mouse release if the mouse cursor is still within the text bounds, and if so, trigger the event. If you still want your behavior (which I really don't think you should, because it's user-unfriendly) you'd have to reimplement that part yourself.

> Also, I placed it in the if dragged condition because all draggable objects would go into that if branch.

This feature is unrelated to drag, even in your use-case, and so should be handled separately.

Reply:
I see what you mean. When typing this comment, if I click down in the middle of the GitHub **Cancel** box, and then drag to the edge - _but still within the box_ - and release, it registers as clicking the box. This is a static setup, though.

Hrm... On the other hand, in a dynamic setup such as a mobile phone's home screen if users click down long enough then they can move the icon... That's another method. Even with my zero-tolerance method, my boxes are easy to click and I find, and expect my customers to find as well, that it is super convenient to be able drag the cluster from anywhere within and it's still super easy to click. [Proof](https://discord.com/channels/338590234235371531/501099842097905675/994993478289330258)

I can't say either method is better than the other. I think it's just a matter of taste/context.

If it's not too much processing overhead and/or too clunky/counterintuitive to provide for the addon authors, we could provide both. Maybe `image:clickable(bool)` that defaults to false if draggable is true, but can be toggled?

Reply:
> On the other hand, browser favorites are similar to my implementation, they just have a healthy tolerance before moving from a click/release to a click/drag scenario.

This is different because there are discrete places to move it to. It's not a continuous space you can move it around in, but only fixed places before and after other elements, so even massive tolerances before registering a click as a drag are fine. But in our case we either have to remove any tolerance, which would result in people occasionally moving a UI object accidentally instead of clicking it (which is not only annoying if it happens in an important situation, such as a battle, but it also destroys the UI layout, which, for people with certain levels of OCD, like me, is an even greater annoyance) or we make the tolerance big enough to tolerate cases like that, but make it significantly harder to fine-adjust UI elements (which, again, for people with certain OCD levels, like me, is a real annoyance).

> Why don't we provide both? We could let draggable objects default to your suggestion - no *left_click* events - but transition to the browser's left_click-if-under-threshold/drag-if-over-threshold behavior if the user adds a `image:clickable(true)` to an already draggable image. Viola, users can have it both ways.

I'm willing to let myself be convinced that this is useful - at this point I just don't see it. I know you seem to want it, though I have to say I still don't understand why, but I think we should ask people in the channel what they think. It would add noise to our API and I just don't see the benefit. Especially since this is something you could add on your own outside of the API.

That said, I'd be more willing to do it without the tolerance, and just let users mess up their UI if they're not careful. At least that way we don't have to have any special handling in our API and we can just blame the addon devs.

Reply:
Well, if we don't add the tolerance-before-dragging to (the new) draggable-and-clickable objects, addon authors won't be able to add that tolerance before dragging to their addons. That is because A) the images.lua is what is forcing all draggable objects to go into drag mode the moment the mouse is clicked down and B) requiring their userbase to modify their local images.luas to use their addons is likely a death wish for any hopes they had for users adopting their addon.

We would need to add that logic where, for the users who did `image:draggable(true)` and then `image:clickable(true)`, the drag event doesn't start on click until the mouse moves out of the tolerance zone. As is, the drag event starts immediately on click.

Reply:
If an author such as myself wanted to add this feature - _a tolerance before the drag event begins on draggable objects, where only a click event fires on release if the tolerance is not exceeded_ - would we be able to do that by packaging our own images.lua with our addon?

Reply:
If an author such as myself wanted to add this feature - _a tolerance before the drag event begins on draggable objects, where only a click event fires on release if the tolerance is not exceeded_ - would we be able to do that by packaging our own images.lua with our addon? I would think that would work.

What I'm wanting, like you said, seems to belong more in a new widgets.lua library, right?